### PR TITLE
MongoDB: More ad hoc fixes for supporting real-world data

### PR DESCRIPTION
**Pull request by amotl**
## About
- Add support for UUID types.
- Improve reading timestamps in previous BSON formats like `end_date: Long('1455141600000')`.
- Make `ctk load table` use the `data OBJECT(DYNAMIC)` mapping strategy.
- Add relevant end-to-end integration test.
- Sanitize lists of varying objects.
- ~~Add `--treatment` option for applying special treatments to certain items on real-world data.~~
- Improve Zyp Transformations (`--transformation` option) for applying special treatments to certain items on real-world data.
- Use pagination on source collection, for creating batches towards CrateDB.
- Unlock importing MongoDB Extended JSON files using `file+bson://...`.

## Documentation
https://cratedb-toolkit--255.org.readthedocs.build/io/mongodb/loader.html

## Install
```
pip install --upgrade 'cratedb-toolkit[mongodb] @ git+https://github.com/crate/cratedb-toolkit.git@mongodb-more'
```

## Backlog
- [x] Add relevant software tests.
- [x] Documentation. For example, parameter `batch-size`.

/cc @hlcianfagna, @hammerhead, @lservini, @zolbatar, @juanpardo, @widmogrod

### CODE REVIEW

File: cratedb_toolkit/io/mongodb/export.py [hunk 1]
```diff
@@ -65,6 +67,9 @@ def extract_value(value, parent_type=None):
     """
     if isinstance(value, dict):
         if len(value) == 1:
+            if "$binary" in value and value["$binary"]["subType"] in ["03", "04"]:
+                decoded = UUID(bytes=base64.b64decode(value["$binary"]["base64"]))
+                return extract_value(decoded, parent_type)
             for k, v in value.items():
                 if k.startswith("$"):
                     return extract_value(v, k.lstrip("$"))
```
Comment (amotl) [inline cratedb_toolkit/io/mongodb/export.py:75]: Sure enough, decoding MongoDB Extended JSON should not be performed manually, but should [use `_json_convert` instead](https://github.com/crate/commons-codec/blob/v0.0.14/src/commons_codec/transform/mongodb.py#L33-L62), like the CDC subsystem is doing it already.

File: cratedb_toolkit/io/mongodb/extract.py [hunk 2]
```diff
@@ -197,4 +198,7 @@ def get_type(value):
             return "INTEGER"
         else:
             return "INT64"
+    if type_ is bson.binary.Binary:
+        if value.subtype in [OLD_UUID_SUBTYPE, UUID_SUBTYPE]:
+            return "UUID"
```
Comment (amotl) [inline cratedb_toolkit/io/mongodb/extract.py:203]: Dito, when applicable.

File: tests/io/mongodb/test_cli.py [hunk 3]
```diff
@@ -29,7 +33,28 @@ def test_version():
...
+DOCUMENT_OUT = {
+    "__id": mock.ANY,
+    "id": "d575540f-759d-4653-a4c4-4a9e410f1aa1",
+    "value": {
+        "name": "foobar",
+        "active": True,
+        "created": 1592579033000,
+        "timestamp": 1455141600000,
+    },
+}
```
Comment (amotl) [inline tests/io/mongodb/test_cli.py:54]: Best view into the integration test added with bf78869b3f943b67b0c0c06e27a00fe54b0c560a.

File: tests/io/mongodb/test_cli.py [hunk 4]
```diff
@@ -29,7 +33,39 @@ def test_version():
...
+CREATE TABLE IF NOT EXISTS "testdrive"."demo" (
+   "__id" TEXT,
+   "id" TEXT,
+   "value" OBJECT(DYNAMIC) AS (
+      "name" TEXT,
+      "active" BOOLEAN,
+      "created" TIMESTAMP WITH TIME ZONE,
+      "timestamp" TIMESTAMP WITH TIME ZONE
+   )
+)""".lstrip()
```
Comment (amotl) [inline tests/io/mongodb/test_cli.py:65]: Best view into the integration test added with bf78869b3f943 and 249cac19ccd020394b.

**Review by wierdvanderhaar — APPROVED** [review]
Lets a go.

File: cratedb_toolkit/io/mongodb/export.py [hunk 5]
```diff
@@ -56,36 +65,157 @@ def timestamp_converter(value):
...
+            value["createdBy"] = {"id": value["createdBy"]}
+
+        # Prune invalid date representations.
+        for key in ["start_date", "end_date"]:
+            if key in value:
+                if not isinstance(value[key], dict):
+                    del value[key]
+                elif "date" in value[key]:
+                    if isinstance(value[key]["date"], str):
+                        del value[key]
```
Comment (amotl) [inline cratedb_toolkit/io/mongodb/export.py:171]: While others are configurable per `--treatment=Path` option using a corresponding YAML file, those special treatments have not been generalized yet.

File: cratedb_toolkit/io/mongodb/export.py (hunk 1 above)
Comment (amotl) [reply cratedb_toolkit/io/mongodb/export.py]: Through `decode_bson`, the machinery now uses BSON's `_json_convert`.

https://github.com/crate/cratedb-toolkit/blob/bc9362be5eab9d183b4b094008de1deb7a8fd00b/cratedb_toolkit/io/mongodb/copy.py#L52-L59

File: cratedb_toolkit/io/mongodb/export.py (hunk 5 above)
Comment (amotl) [reply cratedb_toolkit/io/mongodb/export.py]: Those rules now have been generalized further to be controlled using the treatment file.
```yaml
convert_dict:
  - name: createdBy
    wrapper_name: id
  - name: user
    wrapper_name: id
prune_invalid_date:
  - "start_date"
  - "end_date"
```

Comment (amotl) [reply cratedb_toolkit/io/mongodb/export.py]: The `--treatment` option has been dissolved again. Special treatments have been integrated into Zyp.

- https://github.com/crate/commons-codec/pull/46

File: pyproject.toml [hunk 6]
```diff
@@ -159,12 +161,13 @@ kinesis = [
   "lorrystream[carabas]>=0.0.6",
 ]
 mongodb = [
-  "commons-codec[mongodb,zyp]>=0.0.14",
+  "commons-codec[mongodb,zyp] @ git+https://github.com/crate/commons-codec@zyp-treatments",
```
[the remark this answers is not in the document]
Comment (amotl) [reply pyproject.toml]: commons-codec v0.0.15 has been released.

-- https://pypi.org/project/commons-codec/

**Pull request merged**